ros/wavemap_ros/src/input_handler/pointcloud_undistorter.cc

```diff
@@ -0,0 +1,103 @@
+  const uint64_t step_size =
+      (points.back().time_offset - points.front().time_offset) /
+      (kNumTimeIntervals - 1);
+  const uint64_t buffer_start_time = start_time - step_size;
+  const uint64_t buffer_end_time = end_time + step_size;
+
+  // Make sure all transforms are available
+  if (!transformer_->isTransformAvailable(
+          fixed_frame, stamped_pointcloud.getSensorFrame(),
+          convert::nanoSecondsToRosTime(buffer_start_time))) {
+    return Result::kStartTimeNotInTfBuffer;
+  }
+  if (!transformer_->isTransformAvailable(
+          fixed_frame, stamped_pointcloud.getSensorFrame(),
+          convert::nanoSecondsToRosTime(buffer_end_time))) {
+    return Result::kEndTimeNotInTfBuffer;
+  }
+
+  // Buffer the transforms
+  std::vector<std::pair<uint64_t, Transformation3D>> timed_poses;
+  timed_poses.reserve(kNumTimeSteps);
+  for (unsigned int step_idx = 0u; step_idx < kNumTimeSteps; ++step_idx) {
+    auto& timed_pose = timed_poses.emplace_back();
+    timed_pose.first = start_time + step_idx * step_size;
+    if (!transformer_->lookupTransform(
+            fixed_frame, stamped_pointcloud.getSensorFrame(),
+            convert::nanoSecondsToRosTime(timed_pose.first),
+            timed_pose.second)) {
+      ROS_WARN_STREAM("Failed to buffer intermediate pose at time "
+                      << convert::nanoSecondsToRosTime(timed_pose.first)
+                      << ".");
+      return Result::kIntermediateTimeNotInTfBuffer;
+    }
+  }
+
+  // Motion undistort
+  // NOTE: The undistortion is done by transforming the points into a fixed
+  //       (inertial) frame using the sensor's pose at each point's timestamp.
+  const auto num_points = static_cast<int>(points.size());
+  Eigen::Matrix<FloatingPoint, 3, Eigen::Dynamic> t_W_points;
+  t_W_points.resize(3, num_points);
+  uint64_t previous_point_time = -1u;
+  int pose_left_idx = 0u;
+  Transformation3D T_WCi;
+  Transformation3D T_WCmedian;
+  for (int idx = 0u; idx < num_points; ++idx) {
+    const auto& point = points[idx];
+    const Point3D& Ci_p = point.position;
+
+    // Get the sensor pose at the current point's time stamp
+    const uint64_t time = stamped_pointcloud.getTimeBase() + point.time_offset;
+    if (time != previous_point_time) {
+      previous_point_time = time;
+      while (timed_poses[pose_left_idx + 1].first < time &&
+             pose_left_idx + 2 < kNumTimeSteps) {
+        ++pose_left_idx;
+      }
+      CHECK_LT(pose_left_idx + 1, timed_poses.size());
+      const uint64_t time_left = timed_poses[pose_left_idx].first;
+      const uint64_t time_right = timed_poses[pose_left_idx + 1].first;
+      const Transformation3D& T_WCleft = timed_poses[pose_left_idx].second;
+      const Transformation3D& T_WCright = timed_poses[pose_left_idx + 1].second;
+      FloatingPoint a = static_cast<FloatingPoint>((time - time_left)) /
+                        static_cast<FloatingPoint>((time_right - time_left));
+      CHECK_GE(a, 0.f);
```

Comment:
This is a me vs. the entire maplab team thing, but I reaaaally hate checkfails that could reasonably happen in normal operation. This COULD potentially happen due to very minor floating point precision problems, right? And interpolating to 0.001s in the future wouldn't be a killer issue.

Reply:
This is a leftover from initial testing that I forgot to take out. Will change it to DCHECK so it only triggers if it fails in CI.